Make sure Nodes always have `.phantomChildren`

addaleax opened the pull request:
Fixes a bug (for me) which occurred when npm was calling
`Object.keys()` on the `.phantomChildren` property of a node even
though there was no such property.

I don’t know why the property was not set in that particular case
and I couldn’t reproduce it once I got things working using this
patch; it doesn’t seem like something that would hurt, though.

([npm-debug.log](https://github.com/npm/read-package-tree/files/409083/npm-debug.log.txt) of the failed install, fwiw.)

addaleax committed:
Make sure Nodes always have `.phantomChildren`

Fixes a bug (for me) which occurred when npm was calling
`Object.keys()` on the `.phantomChildren` property of a node even
though there was no such property.

I don’t know why the property was not set in that particular case
and I couldn’t reproduce it once I got things working using this
patch; it doesn’t seem like something that would hurt, though.

othiym23 commented:
I investigated why the Travis run with your patch was failing, and in the process @iarna and I discovered that `read-package-tree`'s tests now fail under Node 6. They work fine under Node 4. :fire::fire::fire: Your change passes under Node 4 just fine, although I'd like a test before merging.

I've only just begun investigating, but if I had to guess right now, I'd say it's probably something related to `glob` / the `fs.realpath` issues in Node 6. If you felt like nailing down whether that's what's happening, that would be super useful. In the meantime, I'm going to make sure we don't have something dicey happening in the tests themselves.

Thanks for putting this together, and thanks for your help, as always!

addaleax commented:
Yup, I’ll look into why this is failing on Node v6 (if I’m understanding you right? :D)

addaleax commented:
Also, yes, the tests seem to be flaky, I guess that’s why I didn’t catch this in the first place…

othiym23 commented:
@iarna has found and fixed the proximal cause of the flakiness in the tests. She'll have a PR up shortly. Unfortunately, the strategy that `read-package-tree` is using to deal with package paths that can be symlink targets is pretty racy, and we're sort of relying on the fact that npm-created package trees are unlikely to be set up in the way that's racy. Also, Rebecca's changes should make the tests reliable, so the only thing blocking this PR from landing is a test to go with this particular change. Thanks!

addaleax committed:
Add test checking the basic layout of a Node instance

addaleax committed:
Make sure directories for symlinks always exist

Adds the missing test/fixtures/badlink/node_modules directory.

As the directory would usually have no contents except during testing,
it isn’t present in a normal git checkout of this repository and so
needs to be created manually in order for the tests to work.

Work around that by letting the test file create it.

addaleax commented:
Rebased & added a line in the test script that creates the parent directories of symlinks before creating the symlinks themselves… I had to create `test/fixtures/badlink/node_modules` myself in order to get green tests locally, and judging from the previous red CI results here, Travis needs the same.

(Sorry for the commit spam if you’re getting notifications, I still haven’t quite figured out when GH sends something and when not?)

iarna commented:
`phantomChildren` isn't a `read-package-tree` construct, it is purely an `npm` one and it should be dealt with there. Adding that here would paper over the problem—it would fix a crash in trying to access `phantomChildren` but leave potential crashes around other defaults.

So between that and the fact that we believe this is entirely fixed in `npm` now, I'm going to close this. Also planned for the mid term is a rewrite of `read-package-tree` to both be substantially faster and to be factored better. (So no initializing the tree in multiple places.)
